### PR TITLE
Templated Field functions and deprecate REGION enum arguments

### PR DESCRIPTION
**Pull request by johnomotani**
Adding `"RGN_NOCORNERS"` in #1719 pushed me to want to be able to use `const std::string& region` arguments everywhere instead of `REGION region`. I think that was the long-term plan anyhow. I've moved most of the functions `abs()`, `sin()`, etc. into `field.hxx` as template functions (using `bout::utils::EnableIfField` to make sure they don't clash with functions of non-`Field` variables - was very glad when I found that in the comments at the top of `field.hxx`! :smile:). There are deprecated overloads with `REGION region` argument for all the changed functions.

This PR reduces the total number of lines of code by a reasonable amount already, and I guess it should be at least a few hundred more once the deprecated forms are deleted. It is a bit big though - might be a good plan to leave off merging until after 4.3, although I think it's ready to go :crossed_fingers:

### CODE REVIEW

File: include/bout/index_derivs_interface.hxx [hunk 1]
```diff
@@ -312,17 +312,17 @@ T VDDY(const T& vel, const T& f, CELL_LOC outloc = CELL_DEFAULT,
...
                                                                     method, region);
   } else {
-    const T f_aligned = toFieldAligned(f, RGN_NOX);
-    const T vel_aligned = toFieldAligned(vel, RGN_NOX);
+    const T f_aligned = toFieldAligned(f, "RGN_NOX");
+    const T vel_aligned = toFieldAligned(vel, "RGN_NOX");
     T result = flowDerivative<T, DIRECTION::Y, DERIV::Upwind>(vel_aligned, f_aligned,
                                                               outloc, method, region);
-    return fromFieldAligned(result, region);
+    return fromFieldAligned(result, toString(region));
```
Comment (ZedThree) [inline include/bout/index_derivs_interface.hxx]: Is `toString` needed here?

File: include/derivs.hxx [hunk 2]
```diff
@@ -50,12 +90,9 @@
...
 ///                    If not given, defaults to RGN_NOBNDRY
-const Field3D DDX(const Field3D& f, CELL_LOC outloc = CELL_DEFAULT,
-                  const std::string& method = "DEFAULT", REGION region = RGN_NOBNDRY);
-inline const Field3D DDX(const Field3D& f, CELL_LOC outloc, DIFF_METHOD method,
-                         REGION region = RGN_NOBNDRY) {
-  return DDX(f, outloc, toString(method), region);
-};
+const Field3D DDX(const Field3D& f, CELL_LOC outloc = CELL_DEFAULT, const std::string&
+    method = "DEFAULT", const std::string& region = "RGN_NOBNDRY");
+DERIV_FUNC_REGION_ENUM_TO_STRING(DDX, Field3D)
```
Comment (ZedThree) [inline include/derivs.hxx]: As we're touching this might, may as well drop the `const` from the return value (similarly in the macro)

**Comment by ZedThree** [comment]
I think you've missed some uses of the deprecated `toFieldAligned` in `fv_ops.hxx` btw

We probably want this in 4.3 so we can remove all the deprecated overloads in 5.0. I'm not sure we can deprecate the `REGION` enum completely right now, as we're using it all over the place in these deprecated functions! We might have to remove these in 5.0 then deprecate `REGION` and then remove that in 5.1

File: include/field.hxx [hunk 3]
```diff
@@ -239,15 +243,15 @@ inline bool areFieldsCompatible(const Field& field1, const Field& field2) {
 
 /// Return an empty shell field of some type derived from Field, with metadata
 /// copied and a data array that is allocated but not initialised.
-template<typename T>
+template<typename T, typename = bout::utils::EnableIfField<T>>
```
[the remark this answers is not in the document]
Comment (johnomotani) [reply include/field.hxx]: :+1: I'll update this PR when we've merged #1730.

File: include/derivs.hxx (hunk 2 above)
Comment (johnomotani) [reply include/derivs.hxx]: Do we want to drop the `const`s in `derivs.cxx` as well?

Comment (johnomotani) [reply include/derivs.hxx]: Ignore that, compiler error tells me we do drop those `const`s too...

File: include/field.hxx [hunk 4]
```diff
@@ -266,7 +270,389 @@ inline T filledFrom(const T& f, BoutReal fill_value) {
...
+}
+template<typename T, typename = bout::utils::EnableIfField<T>>
+[[gnu::deprecated("Please use fromFieldAligned(const T& f, "
+    "const std::string& region = \"RGN_ALL\") instead")]]
+inline T fromFieldAligned(const T& f, REGION region) {
+  return fromFieldAligned(f, toString(region));
+}
+
+template<typename T, typename = bout::utils::EnableIfField<T>>
+inline BoutReal min(const T& f, bool allpe = false, const std::string& rgn = "RGN_NOBNDRY") {
```
[the remark this answers is not in the document]
Comment (johnomotani) [reply include/field.hxx]: I think I've updated this in the appropriate places now. @ZedThree could you check it looks sensible?

File: include/derivs.hxx (hunk 2 above)
Comment (johnomotani) [reply include/derivs.hxx]: I guess we want to do the same in `coordinates.hxx`/`coordinates.cxx` too? I've made that update in the last commit.

**Comment by johnomotani** [comment]
I've updated the deprecated function calls in tests and examples, and fixed a few other compiler warnings while I was there.

File: include/bout/coordinates.hxx [hunk 5]
```diff
@@ -139,101 +139,96 @@ public:
...
   ///////////////////////////////////////////////////////////
 
-  const Field2D DDX(const Field2D& f, CELL_LOC outloc = CELL_DEFAULT,
-                    const std::string& method = "DEFAULT", REGION region = RGN_NOBNDRY);
-  const Field2D DDX(const Field2D& f, CELL_LOC outloc, DIFF_METHOD method,
-                    REGION region = RGN_NOBNDRY) {
+  Field2D DDX(const Field2D& f, CELL_LOC outloc = CELL_DEFAULT,
+      const std::string& method = "DEFAULT", const std::string& region = "RGN_NOBNDRY");
+  Field2D DDX(const Field2D& f, CELL_LOC outloc, DIFF_METHOD method,
+      const std::string& region = "RGN_NOBNDRY") {
```
Comment (ZedThree) [inline include/bout/coordinates.hxx]: We still need to provide the `REGION` overloads, just deprecated, before we remove them after 4.3

File: include/bout/index_derivs_interface.hxx [hunk 6]
```diff
@@ -198,71 +198,71 @@ T D4DX4(const T& f, CELL_LOC outloc = CELL_DEFAULT, const std::string& method =
...
     ASSERT1(f.getDirectionY() == YDirectionType::Standard);
     return standardDerivative<T, DIRECTION::YOrthogonal, DERIV::Standard>(f, outloc,
                                                                           method, region);
   } else {
-    const T f_aligned = toFieldAligned(f, RGN_NOX);
+    const T f_aligned = toFieldAligned(f, "RGN_NOX");
     T result = standardDerivative<T, DIRECTION::Y, DERIV::Standard>(f_aligned, outloc,
                                                                     method, region);
-    return fromFieldAligned(result, region);
+    return fromFieldAligned(result, toString(region));
```
Comment (ZedThree) [inline include/bout/index_derivs_interface.hxx]: `region` is already a `string` here :) Similarly in the next few functions

**Comment by ZedThree** [comment]
PR test is failing because next was failing. Restarted now that next is fixed, though I can't remember if Travis will use the updated next. We'll find out!

Also, for reference, this removes `REGION` from the derivative factory stuff entirely. It wasn't immediately obvious how one might keep a deprecated `REGION` version around at the same time as the `string` without massive code duplication.

**Pull request merged**